### PR TITLE
#25 Implement additional Package/NamingRule

### LINKED ISSUE #25: Implement additional Package/NamingRule
This story is an extension of story #6 with PR #12 such that classes/interfaces extending:
* `ApplicationPersistenceEntity` have to have be in layer `dataaccess` (already implemented that they have to end with `Entity`)
* `JpaRepository` have to be in layer `dataaccess` and have to be named `«EntityName»Repository` where `«EntityName»` is the name of the entity filled in the generic argument of `JpaRepository` excluding the `Entity` suffix. Further they should be in the same package as the entity.
* `AbstractEto` have to be in layer `common` (already implemented that they have to end with `Eto`) and additionally they implement an interface with the same simple-name excluding the `Eto` suffix. Also that interface shall be in the same package.
* `AbstractUc` have to be in layer `logic` (already implemented that they have to start with `Uc`).

Finally classes annotated with 
* `@Mapper` (org.mapstruct.Mapper) need to be in layer `logic` and have to end with `Mapper`.
* `@Path` (javax.ws.rs.Path)  need to be in layer `service` and have to end with `Service`.

### PR DESCRIPTION
**Pull request by moritzLanger**

### CODE REVIEW

File: src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java [hunk 1]
```diff
@@ -37,5 +49,88 @@ public class NamingConventionTest {
...
                     .should().haveSimpleNameEndingWith("Eto")
                     .because("Classes extending AbstractEto must follow the naming convention by ending with 'Eto'.");
+    /**
+     * DevonApplicationPersistenceEntityCheck verifying that classes extending ApplicationPersistenceEntity have to have be in layer dataaccess
+     */
+    @ArchTest
+    private static final ArchRule DevonApplicationPersistenceEntityCheck = 
+            classes()
+                   .that().areAssignableTo(ApplicationPersistenceEntity.class).should().resideInAnyPackage("..dataaccess..")
+                   .because("Classes extending ApplicationPersistenceEntity must be located in the package dataaccess");
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java]: I would merge this together with the existing Entity rule `N3DevonNamingConventionEntityCheck`.

File: src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java [hunk 2]
```diff
@@ -37,5 +49,88 @@ public class NamingConventionTest {
...
+                   {
+                        @Override  
+                        public void check(JavaClass javaClass, ConditionEvents events) {
+                                String supposedInterfaceName = javaClass.getPackageName() + "." + javaClass.getSimpleName().replace("Eto", "");
+                                boolean hasCorrectInterface = javaClass.getInterfaces().stream().anyMatch(i->i.getName().equals(supposedInterfaceName));
+                                String message = "The Testresult of " + javaClass.getSimpleName() + " was " + hasCorrectInterface;
+                                events.add(new SimpleConditionEvent(javaClass, hasCorrectInterface, message));
+                        }
+                        
+                   });
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java]: I would merge this with the existing ETO rule in lines 42-51.

File: src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java [hunk 3]
```diff
@@ -37,5 +49,88 @@ public class NamingConventionTest {
...
+       
+      /**
+     * DevonJpaRepositoryCheck verifying that classes implementing JpaRepository have to have be in layer dataaccess 
+     * and have to be named «EntityName»Repository where «EntityName» is the name of the entity filled in the generic argument of JpaRepository excluding the Entity suffix. 
+     * Further they should be in the same package as the entity.
+     */               
+    @ArchTest
+    private static final ArchRule DevonJpaRepositoryCheck = 
+            classes().that().areAssignableTo(JpaRepository.class)
+                  .should().resideInAnyPackage("..dataaccess..")
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java]: `.andShould().haveSimpleNameEndingWith("Repository")`

File: src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java [hunk 4]
```diff
@@ -37,5 +49,88 @@ public class NamingConventionTest {
...
+                   .because("Classes extending AbstractUc must be located in the package logic");
+       
+     /**
+     * DevonMapperCheck verifying that classes extending Mapper have to have be in layer logic and end with the suffix Mapper
+     */
+    @ArchTest
+    private static final ArchRule DevonMapperCheck = 
+            classes()   
+                   .that().areAnnotatedWith(Mapper.class).should().resideInAnyPackage("..logic..").andShould().haveSimpleNameEndingWith("Mapper")
+                   .because("Classes extending Mapper must be located in the package logic and end with Mapper");
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java]: replace `Classes extending Mapper` with `Bean-Mappers` or `Types annotated with @Mapper` as our mappers are interfaces and not classes and they do not extend `Mapper` but are annotated with it.

File: src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java [hunk 5]
```diff
@@ -37,5 +49,88 @@ public class NamingConventionTest {
...
+
+     /**
+     * DevonPathCheck verifying that classes extending Path have to have be in layer service and end with the suffix Service
+     */
+    @ArchTest
+    private static final ArchRule DevonPathCheck = 
+            classes()   
+                   .that().areAnnotatedWith(Path.class).should().resideInAnyPackage("..service..").andShould().haveSimpleNameEndingWith("Service")
+                   .because("Classes extending Path must be located in the package service and end with Service");
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java]: replace `Classes extending Path` with `Rest-services` or `Types annotated with @Path`.

File: src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java [hunk 6]
```diff
@@ -51,6 +63,90 @@ public class NamingConventionTest {
...
+                   {
+                        @Override  
+                        public void check(JavaClass javaClass, ConditionEvents events) {
+                                String supposedInterfaceName = javaClass.getPackageName() + "." + javaClass.getSimpleName().replace("Eto", "");
+                                boolean hasCorrectInterface = javaClass.getInterfaces().stream().anyMatch(i->i.getName().equals(supposedInterfaceName));
+                                String message = "The Testresult of " + javaClass.getSimpleName() + " was " + hasCorrectInterface;
+                                events.add(new SimpleConditionEvent(javaClass, hasCorrectInterface, message));
+                        }
+                        
+                   });
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java]: I would merge this with the existing ETO rule `N4DevonNamingConventionEtoCheck`.

File: src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java [hunk 7]
```diff
@@ -51,6 +63,90 @@ public class NamingConventionTest {
...
+                   .that().areAssignableTo(ApplicationPersistenceEntity.class).should().resideInAnyPackage("..dataaccess..")
+                   .because("Classes extending ApplicationPersistenceEntity must be located in the package dataaccess");
+
+    /**
+     * DevonAbstractUcCheck verifying that classes extending AbstractUc have to have be in layer logic
+     */
+    @ArchTest
+    private static final ArchRule DevonAbstractUcCheck = 
+            classes()
+                   .that().areAssignableTo(AbstractUc.class).should().resideInAnyPackage("..logic..")
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java]: `.andShould().haveSimpleNameStartingWith("Uc")`

ATTENTION: We do have some base-usecases starting with `AbstractUc` so either we add this with OR conjunction or even better we distinguish abstract from non-abstract classes here.

File: src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java [hunk 8]
```diff
@@ -51,6 +63,90 @@ public class NamingConventionTest {
...
+                   .because("Classes extending AbstractUc must be located in the package logic");
+       
+    /**
+     * DevonMapperCheck verifying that classes extending Mapper have to have be in layer logic and end with the suffix Mapper
+     */
+    @ArchTest
+    private static final ArchRule DevonMapperCheck = 
+            classes()
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java]: does `classes()` cover all types or just classes?
Please note that our mappers and repositories (see line 131) are interfaces.

**Review by hohwille — CHANGES_REQUESTED** [review]
@moritzLanger thanks for your PR. Great job, you have implemented the new rules well 👍 
Still I have some review comments to align new with existing rules to make this more readable and have all requirements in one place. Also I found some small things to improve. Please have a look at my review comments.

File: src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java (hunk 8 above)
Comment (moritzLanger) [reply src/test/java/com/devonfw/sample/archunit/NamingConventionTest.java]: Yes in Archunit Interfaces are included in the methods classes() and noclasses().

**Review by hohwille — APPROVED** [review]
@moritzLanger Thanks for your rework and all the improvements addressing my review comments. 👍 

We still have two rules for `AbstractUc` spread across the test, but I will simply merge and this can still be improved once there is some time left.

**Pull request merged**